Documentation/contributing/testing/bpf.rst

```diff
@@ -0,0 +1,307 @@
+a name it serves to self-document, and makes it easier to spot what part of a test fails.
+
+Integration tests
+-----------------
+
+Writing tests for a single function or small group of functions should be fairly straightforward, 
+only requiring a ``CHECK`` program. Testing functionality across tail calls requires an additional step,
+Since we will not return to the ``CHECK`` function after making a tail call, we can't check if it was
+successful.
+
+The fix is to use a ``SETUP`` program in addition to a ``CHECK`` program. ``SETUP`` programs with the same
+name will run before the ``CHECK`` program. The context, modified by ``SETUP`` is then passed to ``CHECK``
+which can inspect the result. By executing the test setup and executing the tail call in ``SETUP`` 
+we can execute complete programs. The return code of the ``SETUP`` program is prefixed as ``u32`` to
+the start of the packet data passed to ``CHECK``. 
+
+This is a abbreviated example showing the key components:
+
+.. code-block:: c
+    
+    #include "common.h"
+
+    #include "bpf/ctx/xdp.h"
+    #include "bpf_xdp.c"
+
+    struct {
+        __uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+        __uint(key_size, sizeof(__u32));
+        __uint(max_entries, 2);
+        __array(values, int());
+    } entry_call_map __section(".maps") = {
+        .values = {
+            [0] = &bpf_xdp_entry,
+        },
+    };
+
+    SETUP("xdp", "l2_example")
+    int test1_setup(struct __ctx_buff *ctx)
+    {
+        /* Create room for our packet to be crafted */
+        unsigned int data_len = ctx->data_end - ctx->data;
+        int offset = offset = sizeof(struct ethhdr) - data_len;
+        bpf_xdp_adjust_tail(ctx, offset);
+
+        void *data = (void *)(long)ctx->data;
+        void *data_end = (void *)(long)ctx->data_end;
+
+        if (data + sizeof(struct ethhdr) > data_end)
+            return TEST_ERROR;
+
+        /* Writing just the L2 header for brevity */
+        struct ethhdr l2 = {
+            .h_source = {0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF},
+            .h_dest = {0x12, 0x23, 0x34, 0x45, 0x56, 0x67},
+            .h_proto = bpf_htons(ETH_P_IP)
+        };
+        memcpy(data, &l2, sizeof(struct ethhdr));
+       
+        /* OMITTED setting up map state */
+
+        /* Jump into the entrypoint */
+        tail_call_static(ctx, &entry_call_map, 0);
+        /* Fail if we didn't jump */
+        return TEST_ERROR;
+    }
+
+    CHECK("xdp", "l2_example")
+    int test1_check(__maybe_unused const struct __ctx_buff *ctx)
+    {
+        test_init();
+
+        void *data = (void *)(long)ctx->data;
+        void *data_end = (void *)(long)ctx->data_end;
+
+        if (data + sizeof(__u32) > data_end)
+            test_fatal("status code out of bounds");
+
+        __u32 *status_code = data;
+
+        if (*status_code != XDP_TX)
+            test_fatal("status code != XDP_TX");
+
+        data += sizeof(__u32);
```

Comment:
All of this could maybe go into a helper function too? `setup_assert_status(XDP_TX)` which would also add the offset to `data`? I don't know